mobile/lib/modules/backup/services/background.service.dart

```diff
@@ -0,0 +1,147 @@
+    bool immediate = false,
+    bool keepExisting = false,
+    bool requireUnmetered = true,
+    bool requireCharging = false,
+  }) async {
+    if (!_isForegroundInitialized) {
+      await _initialize();
+    }
+    debugPrint("Starting background service...");
+    return await _foregroundChannel.invokeMethod(
+        'start', [immediate, keepExisting, requireUnmetered, requireCharging]);
+  }
+
+  static Future<bool> stopService() async {
+    if (!_isForegroundInitialized) {
+      await _initialize();
+    }
+    debugPrint("Stopping background service...");
+    return await _foregroundChannel.invokeMethod('stop');
+  }
+
+  static Future<bool> updateNotification({
+    String title = "Immich",
+    String? content,
+  }) async {
+    if (_isBackgroundInitialized) {
+      return await _backgroundChannel
+          .invokeMethod('updateNotification', [title, content]);
+    }
+    return Future.value(false);
+  }
+
+  static Future<bool> showErrorNotification(
+      String title, String content) async {
+    if (_isBackgroundInitialized) {
+      return await _backgroundChannel
+          .invokeMethod('showError', [title, content]);
+    }
+    return Future.value(false);
+  }
+
+  static void _setupBackgroundCallHandler() {
+    _backgroundChannel.setMethodCallHandler(_callHandler);
+    _isBackgroundInitialized = true;
+    _backgroundChannel.invokeMethod('initialized');
+  }
+
+  static Future<bool> _callHandler(MethodCall call) async {
+    switch (call.method) {
+      case "onPhotosChanged":
+        final Map<String, List<Object?>> data =
+            call.arguments.cast<String, List<Object?>>();
+
+        final String text = data["urls"].toString();
+        debugPrint("received $text");
+
+        try {
+          await _onPhotosChanged(data["urls"]!.cast<String>());
+          return Future.value(true);
+        } catch (error) {
+          debugPrint(error.toString());
+          return Future.value(false);
+        }
+
+      default:
+        debugPrint("Unknown method ${call.method}");
+        return Future.value(false);
+    }
+  }
+}
+
+/// Needs to be a top-level function
+void _callbackDispatcher() {
+  WidgetsFlutterBinding.ensureInitialized();
+  BackgroundService._setupBackgroundCallHandler();
+}
+
+/// Main method for dart/flutter immich when running as a background service
+Future<void> _onPhotosChanged(List<String> changedUrls) async {
+  final container = ProviderContainer();
+  await Hive.initFlutter();
+
+  Hive.registerAdapter(HiveSavedLoginInfoAdapter());
+  Hive.registerAdapter(HiveBackupAlbumsAdapter());
+
+  await Hive.openBox(userInfoBox);
+  await Hive.openBox<HiveSavedLoginInfo>(hiveLoginInfoBox);
+  await Hive.openBox<HiveBackupAlbums>(hiveBackupInfoBox);
+
+  HiveSavedLoginInfo? loginInfo =
+      Hive.box<HiveSavedLoginInfo>(hiveLoginInfoBox).get(savedLoginInfoKey);
+
+  var isAuthenticated =
+      await container.read(authenticationProvider.notifier).login(
```

Comment:
Maybe we don't need to use riverpod in the background service since Riverpod is for state management. We can use the generated OpenAPI SDK to make request directly to the server.

Reply:
You can take a look at `ApiService` class at `/mobile/lib/shared/services/api.services.dart`

Reply:
that could be an option. Could lead to some code duplication (backup/uploading in app foreground vs service would take separate paths?).

Reply:
This is a fair point. Let's keep the current backup code flow

Reply:
keeping the current backup code contradicts with the following, or does it not?

> 5. How to make the background service as lean as possible

> With the current ApiService, we can do this very easily.

Reply:
I'm not familiar with the app code at all - can you not just call into the existing backup code? (Or refactor it a bit to make that possible)

Reply:
it's possible, that's currently what this PR does. but it has its issues ([see 5.](https://github.com/immich-app/immich/pull/382#issuecomment-1199445946))